src/utils/Petsc.hpp

```diff
@@ -62,24 +62,40 @@ class Vector
   /// Creates a new vector on the given MPI communicator.
   explicit Vector(std::string name = "");
 
-  /// Use Vec v as vector.
-  Vector(Vec &v, std::string name = "");
+  /** Copy construction from another vector
+   * Duplicates the vector and copies the name
+   */
+  Vector(const Vector& other);
+
+  /** Copy assignement
+   * Destroys the current vector and takes ownership of the other.
+   */
+  Vector& operator=(Vector other);
+
+  /** Move construction
+   * Takes ownership of the other vector.
+   */
+  Vector(Vector&& other);
+
+  /** Move assignement
+   * Destroys the current vector and takes ownership of the other vector.
+   */
+  Vector& operator=(Vector&& other);
 
   /// Duplicates type, row layout etc. (not values) of v.
-  Vector(Vector &v, std::string name = "");  
+  Vector(Vector &v, std::string name);  
```

Comment:
As the comment correctly states this does not (deep) copy the vector. Therefore it is not equivalent to the copy constructor. I would suggest:
```suggestion
  explicit Vector(Vector &v, std::string name = "");  
```

Reply:
Good idea!
We should mark the other ctors explicit too, as they all are candidates for implicit conversions. 

So, making this ctor explicit would remove the semantic compatability with the copy ctor.

However, this "alias" functionality clearly breaks the model of unique ownership wrt the underlying vector.
So, either we go all-in on shared ownership and make them all owners or we remove this alias functionality.

Reply:
This is no alias. This is just saying "I have a vector here. Create a similar one for me". No pointers are ever copied (the [definition](https://github.com/precice/precice/blob/f71f6e4d0fe29811aaaee04f28d6385067be0277/src/utils/Petsc.cpp#L142)). 
I would also make this take a `const Vector& v` and cast away `const` inside the c`tor.

Reply:
Oh I get it now.

In that case, I argue that this syntax is too close to copy construction.

We could either
1) replace this functionality into a factory function  
`Vector::makeSimilar(Vector& other, string name)`, or
2) add a tag to the constructor  
`Vector(allocate_only_tag, Vector &v, std::string name)`.  
This makes it difficult to keep the constructors consistent in the long run.

I would definitely go for the first option as it make clear that we ask for allocation and not a copy.

Reply:
Sounds good.

Reply:
The ctor is not covered by the [tests](https://codecov.io/gh/precice/precice/src/c428c7f3e791bd39fbdf1507ccc1d500e92c3fb4/src/utils/Petsc.cpp#L108...118) and compiling with `[[deprecated]]` did not raise a warning.

So I presume this is actually dead code.